### ISSUE TITLE
Pyface Optimizing error.

### ISSUE THREAD

**Issue by 7369632** [Nuitka/Nuitka#376]
python Version:Python2.7.16 32bit
pyface==6.0.0
mayavi==4.6.2
VTK==7.1.1

Error Report:
```
Nuitka:INFO:Optimizing module 'pyface.tree.tree_model', 117 more modules to go after that.Memory usage 2.25 GB (2419085312 bytes):
Nuitka:INFO:Doing module local optimizations for 'pyface.tree.tree_model'.
Problem with statement at D:\8-1Pack\Anaconda\lib\site-packages\pyface\tree\tree_model.py:97:
->     def get_key(self, node):

Problem with statement at D:\8-1Pack\Anaconda\lib\site-packages\pyface\tree\tree_model.py:97:
->     def get_key(self, node):

Problem with statement at D:\8-1Pack\Anaconda\lib\site-packages\pyface\tree\tree_model.py:24:
-> class TreeModel(HasTraits):

Problem with statement at D:\8-1Pack\Anaconda\lib\site-packages\pyface\tree\tree_model.py:24:
-> class TreeModel(HasTraits):

Problem with statement at D:\8-1Pack\Anaconda\lib\site-packages\pyface\tree\tree_model.py:24:
-> class TreeModel(HasTraits):

Nuitka:INFO:Interrupted while working on '<Node 'COMPILED_PYTHON_MODULE' with {'package': 'pyface.tree', 'name': 'tree_model', 'filename': 'D:\\8-1Pack\\Anaconda\\lib\\site-packages\\pyface\\tree\\tree_model.py'}>'.
Traceback (most recent call last):
  File "d:\8-1pack\anaconda\lib\site-packages\Nuitka-0.6.4rc3-py2.7.egg\nuitka\__main__.py", line 184, in <module>
    main()
  File "d:\8-1pack\anaconda\lib\site-packages\Nuitka-0.6.4rc3-py2.7.egg\nuitka\__main__.py", line 177, in main
    MainControl.main()
  File "D:\8-1Pack\Anaconda\lib\site-packages\nuitka-0.6.4rc3-py2.7.egg\nuitka\MainControl.py", line 737, in main
    main_module = createNodeTree(filename=filename)
  File "D:\8-1Pack\Anaconda\lib\site-packages\nuitka-0.6.4rc3-py2.7.egg\nuitka\MainControl.py", line 141, in createNodeTree
    Optimization.optimize(main_module.getOutputFilename())
  File "D:\8-1Pack\Anaconda\lib\site-packages\nuitka-0.6.4rc3-py2.7.egg\nuitka\optimizations\Optimization.py", line 561, in optimize
    finished = makeOptimizationPass(initial_pass=False)
  File "D:\8-1Pack\Anaconda\lib\site-packages\nuitka-0.6.4rc3-py2.7.egg\nuitka\optimizations\Optimization.py", line 456, in makeOptimizationPass
    changed = optimizeModule(current_module)
  File "D:\8-1Pack\Anaconda\lib\site-packages\nuitka-0.6.4rc3-py2.7.egg\nuitka\optimizations\Optimization.py", line 167, in optimizeModule
    changed = optimizeCompiledPythonModule(module)
  File "D:\8-1Pack\Anaconda\lib\site-packages\nuitka-0.6.4rc3-py2.7.egg\nuitka\optimizations\Optimization.py", line 100, in optimizeCompiledPythonModule
    module.computeModule()
  File "D:\8-1Pack\Anaconda\lib\site-packages\nuitka-0.6.4rc3-py2.7.egg\nuitka\nodes\ModuleNodes.py", line 489, in computeModule
    trace_collection=self.trace_collection
  File "D:\8-1Pack\Anaconda\lib\site-packages\nuitka-0.6.4rc3-py2.7.egg\nuitka\nodes\StatementNodes.py", line 165, in computeStatementsSequence
    new_statement = statement.computeStatementsSequence(trace_collection)
  File "D:\8-1Pack\Anaconda\lib\site-packages\nuitka-0.6.4rc3-py2.7.egg\nuitka\nodes\FrameNodes.py", line 182, in computeStatementsSequence
    new_statement = trace_collection.onStatement(statement=statement)
  File "D:\8-1Pack\Anaconda\lib\site-packages\nuitka-0.6.4rc3-py2.7.egg\nuitka\optimizations\TraceCollections.py", line 550, in onStatement
    new_statement, change_tags, change_desc = statement.computeStatement(self)
  File "D:\8-1Pack\Anaconda\lib\site-packages\nuitka-0.6.4rc3-py2.7.egg\nuitka\nodes\TryNodes.py", line 117, in computeStatement
    result = tried.computeStatementsSequence(trace_collection=trace_collection)
  File "D:\8-1Pack\Anaconda\lib\site-packages\nuitka-0.6.4rc3-py2.7.egg\nuitka\nodes\StatementNodes.py", line 167, in computeStatementsSequence
    new_statement = trace_collection.onStatement(statement=statement)
  File "D:\8-1Pack\Anaconda\lib\site-packages\nuitka-0.6.4rc3-py2.7.egg\nuitka\optimizations\TraceCollections.py", line 550, in onStatement
    new_statement, change_tags, change_desc = statement.computeStatement(self)
  File "D:\8-1Pack\Anaconda\lib\site-packages\nuitka-0.6.4rc3-py2.7.egg\nuitka\nodes\AssignNodes.py", line 267, in computeStatement
    trace_collection.onExpression(self.getAssignSource())
  File "D:\8-1Pack\Anaconda\lib\site-packages\nuitka-0.6.4rc3-py2.7.egg\nuitka\optimizations\TraceCollections.py", line 531, in onExpression
    r = expression.computeExpressionRaw(trace_collection=self)
  File "D:\8-1Pack\Anaconda\lib\site-packages\nuitka-0.6.4rc3-py2.7.egg\nuitka\nodes\OutlineNodes.py", line 224, in computeExpressionRaw
    result = body.computeStatementsSequence(trace_collection=trace_collection)
  File "D:\8-1Pack\Anaconda\lib\site-packages\nuitka-0.6.4rc3-py2.7.egg\nuitka\nodes\StatementNodes.py", line 167, in computeStatementsSequence
    new_statement = trace_collection.onStatement(statement=statement)
  File "D:\8-1Pack\Anaconda\lib\site-packages\nuitka-0.6.4rc3-py2.7.egg\nuitka\optimizations\TraceCollections.py", line 550, in onStatement
    new_statement, change_tags, change_desc = statement.computeStatement(self)
  File "D:\8-1Pack\Anaconda\lib\site-packages\nuitka-0.6.4rc3-py2.7.egg\nuitka\nodes\TryNodes.py", line 117, in computeStatement
    result = tried.computeStatementsSequence(trace_collection=trace_collection)
  File "D:\8-1Pack\Anaconda\lib\site-packages\nuitka-0.6.4rc3-py2.7.egg\nuitka\nodes\StatementNodes.py", line 167, in computeStatementsSequence
    new_statement = trace_collection.onStatement(statement=statement)
  File "D:\8-1Pack\Anaconda\lib\site-packages\nuitka-0.6.4rc3-py2.7.egg\nuitka\optimizations\TraceCollections.py", line 550, in onStatement
    new_statement, change_tags, change_desc = statement.computeStatement(self)
  File "D:\8-1Pack\Anaconda\lib\site-packages\nuitka-0.6.4rc3-py2.7.egg\nuitka\nodes\LocalsDictNodes.py", line 392, in computeStatement
    trace_collection=trace_collection
  File "D:\8-1Pack\Anaconda\lib\site-packages\nuitka-0.6.4rc3-py2.7.egg\nuitka\nodes\NodeBases.py", line 868, in computeStatementSubExpressions
    expression = trace_collection.onExpression(expression=expression)
  File "D:\8-1Pack\Anaconda\lib\site-packages\nuitka-0.6.4rc3-py2.7.egg\nuitka\optimizations\TraceCollections.py", line 531, in onExpression
    r = expression.computeExpressionRaw(trace_collection=self)
  File "D:\8-1Pack\Anaconda\lib\site-packages\nuitka-0.6.4rc3-py2.7.egg\nuitka\nodes\ExpressionBases.py", line 980, in computeExpressionRaw
    expression = trace_collection.onExpression(sub_expression)
  File "D:\8-1Pack\Anaconda\lib\site-packages\nuitka-0.6.4rc3-py2.7.egg\nuitka\optimizations\TraceCollections.py", line 531, in onExpression
    r = expression.computeExpressionRaw(trace_collection=self)
  File "D:\8-1Pack\Anaconda\lib\site-packages\nuitka-0.6.4rc3-py2.7.egg\nuitka\nodes\FunctionNodes.py", line 830, in computeExpressionRaw
    function_body.computeFunctionRaw(trace_collection)
  File "D:\8-1Pack\Anaconda\lib\site-packages\nuitka-0.6.4rc3-py2.7.egg\nuitka\nodes\FunctionNodes.py", line 384, in computeFunctionRaw
    self.computeFunction(trace_collection)
  File "D:\8-1Pack\Anaconda\lib\site-packages\nuitka-0.6.4rc3-py2.7.egg\nuitka\nodes\FunctionNodes.py", line 397, in computeFunction
    trace_collection=trace_collection
  File "D:\8-1Pack\Anaconda\lib\site-packages\nuitka-0.6.4rc3-py2.7.egg\nuitka\nodes\StatementNodes.py", line 167, in computeStatementsSequence
    new_statement = trace_collection.onStatement(statement=statement)
  File "D:\8-1Pack\Anaconda\lib\site-packages\nuitka-0.6.4rc3-py2.7.egg\nuitka\optimizations\TraceCollections.py", line 550, in onStatement
    new_statement, change_tags, change_desc = statement.computeStatement(self)
  File "D:\8-1Pack\Anaconda\lib\site-packages\nuitka-0.6.4rc3-py2.7.egg\nuitka\nodes\TryNodes.py", line 157, in computeStatement
    statement,
nuitka.Errors.NuitkaOptimizationError: 
This statement does raise but didn't annotate an exception exit.

    <node kind="StatementsFrameFunction" line="97" co_argcount="2" co_kwonlyargcount="0" co_name="get_key" co_new_locals="True" co_filename="pyface\tree\tree_model.py" co_is_optimized="True" co_has_starlist="False" code_flags="" co_has_closure="False" co_lineno="97" co_kind="Function" co_has_stardict="False" co_varnames="self,node,key">
      <role name="statements" type="list">
        <node kind="StatementTry" line="100">
          <role name="tried">
            <node kind="StatementsSequence" line="100">
              <role name="statements" type="list">
                <node kind="StatementAssignmentVariable" line="101" owner="pyface$tree$tree_model$$$function_7_get_key" is_temp="False" variable_name="key">
                  <role name="source">
                    <node kind="ExpressionBuiltinHash" line="101">
                      <role name="value">
                        <node kind="ExpressionVariableRef" line="101" owner="pyface$tree$tree_model$$$function_7_get_key" variable_name="node"/>
                      </role>
                    </node>
                  </role>
                </node>
              </role>
            </node>
          </role>
          <role name="except_handler">
            <node kind="StatementsSequence" line="100">
              <role name="statements" type="list">
                <node kind="StatementPreserveFrameException" line="100" preserver_id="0"/>
                <node kind="StatementPublishException" line="100"/>
                <node kind="StatementAssignmentVariable" line="104" owner="pyface$tree$tree_model$$$function_7_get_key" is_temp="False" variable_name="key">
                  <role name="source">
                    <node kind="ExpressionBuiltinId" line="104">
                      <role name="value">
                        <node kind="ExpressionVariableRef" line="104" owner="pyface$tree$tree_model$$$function_7_get_key" variable_name="node"/>
                      </role>
                    </node>
                  </role>
                </node>
              </role>
            </node>
          </role>
          <role name="break_handler" type="none"/>
          <role name="continue_handler" type="none"/>
          <role name="return_handler" type="none"/>
        </node>
      </role>
    </node>



The above information should be included in a bug report.
```

**Comment by kayhayen** [comment]
What Nuitka version is this?

**Comment by kayhayen** [comment]
Ok, just noticed in the trace back, it's develop version. Nevermind.

**Comment by kayhayen** [comment]
I managed to reproduce it, simply by compiling `./Lib/site-packages/pyface/tree/tree_model.py` from my Anaconda2 installation after `conda install pytree`.

**Comment by kayhayen** [comment]
So this is the minimal reproducer:

```
def get_key(node):
    try:
        key = hash(node)

    except:
        key = id(node)

    return key
```

The `id` node was not annotated to not raise, therefore giving:

This statement does raise but didn't annotate an exception exit.

```
    <node is_temp="False" kind="StatementAssignmentVariable" line="7" owner="__main__$$$function_1_get_key" variable_name="key">
      <role name="source">
        <node kind="ExpressionBuiltinId" line="7">
          <role name="value">
            <node kind="ExpressionVariableRef" line="7" owner="__main__$$$function_1_get_key" variable_name="node" />
          </role>
        </node>
      </role>
    </node>
```

Adding the indication that `id` does not raise an exception ever, allows for better optimization of it, and fixes this issue, where Nuitka found an internal contradiction (does raise or not). 

I am currently testing a fix internally, expect this to show up soon.

Yours,
Kay

**Comment by 7369632** [comment]
I try it now.

**Comment by 7369632** [comment]
> 
> 
> What Nuitka version is this?

Version is Nuitka-0.6.4rc3

**Comment by 7369632** [comment]
Excellent,wish you Good Luck.

**Comment by kayhayen** [comment]
So this is on the factory branch right now. I expect to make a pre-release of this during the week.

**Comment by kayhayen** [comment]
if you want to try it, @7369632 go there: http://nuitka.net/doc/factory.html

**Comment by 7369632** [comment]
> 
> 
> if you want to try it, @7369632 go there: http://nuitka.net/doc/factory.html

No problems found when using.
Thanks.

**Comment by kayhayen** [comment]
This is on develop now.

**Comment by kayhayen** [comment]
Released as 0.6.4
